# SQL vs Pandas

This notebook explores how the sql queries equivalent operations can be performed on data using pandas on tips dataset.

## Table of Contents
- [Sql Select](#select)
- [Sql Where](#where)
- [IS NULL and NOT NULL](#null)
- [Sql IN](#in)
- [Sql Groupby](#groupby)
- [Sql Having](#having)
- [Sql Join](#join)
    - Inner Join
    - Left Join 
    - Right Join
    - Outer/Full Join
    - Joining on Multiple keys
- [Sql Update](#update)
- [Sql Delete](#delete)
- [Sql Union All and Union](#union)
- [Sql Window functions](#windows_functions)

Reference Links:
    - pandas doc for sql: https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_sql.html
    - pandas transform and filter: https://www.drawingfromdata.com/filter-transform-group-with-pandas   
    - SQL Window Functions: https://www.sqltutorial.org/sql-window-functions/
    - goroupby vs partition function : https://www.sqltutorial.org/sql-window-functions/sql-partition-by/

In [1]:
import pandas as pd
import numpy as np 

In [2]:
url = 'https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/tips.csv'
tips = pd.read_csv(url)

tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [3]:
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


In [4]:
tips.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

<a id='select'></a>
## Sql Select 

- Select * from tips
- Select * from tips limit 10
- Select few columns only
- calculate an additional column during select (**assign()** returns all columns in addition to the columns calculated within assign)
- Add a new calculated column in the data
    - using assign 
    - other syntax

In [5]:
'''
Select * from tips
'''
tips

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [6]:
'''
select * from tips limit 10 
'''
tips.head(10)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
6        8.77  2.00    Male     No  Sun  Dinner     2
7       26.88  3.12    Male     No  Sun  Dinner     4
8       15.04  1.96    Male     No  Sun  Dinner     2
9       14.78  3.23    Male     No  Sun  Dinner     2

In [7]:
'''
Select total_bill, tip, time from tips limit 10
'''

tips[['total_bill', 'tip', 'time']].head(10)

total_bill   tip    time
0       16.99  1.01  Dinner
1       10.34  1.66  Dinner
2       21.01  3.50  Dinner
3       23.68  3.31  Dinner
4       24.59  3.61  Dinner
5       25.29  4.71  Dinner
6        8.77  2.00  Dinner
7       26.88  3.12  Dinner
8       15.04  1.96  Dinner
9       14.78  3.23  Dinner

In [8]:
'''
Select *, tip/total_bill as tip_percentage from tips limit 10
'''

tips.assign(tip_percentage= tips['tip']/tips['total_bill']).head(10)

total_bill   tip     sex smoker  day    time  size  tip_percentage
0       16.99  1.01  Female     No  Sun  Dinner     2        0.059447
1       10.34  1.66    Male     No  Sun  Dinner     3        0.160542
2       21.01  3.50    Male     No  Sun  Dinner     3        0.166587
3       23.68  3.31    Male     No  Sun  Dinner     2        0.139780
4       24.59  3.61  Female     No  Sun  Dinner     4        0.146808
5       25.29  4.71    Male     No  Sun  Dinner     4        0.186240
6        8.77  2.00    Male     No  Sun  Dinner     2        0.228050
7       26.88  3.12    Male     No  Sun  Dinner     4        0.116071
8       15.04  1.96    Male     No  Sun  Dinner     2        0.130319
9       14.78  3.23    Male     No  Sun  Dinner     2        0.218539

In [9]:
tips.assign(tip_percent = (tips['tip']/tips['total_bill'])*100).head(5)

total_bill   tip     sex smoker  day    time  size  tip_percent
0       16.99  1.01  Female     No  Sun  Dinner     2     5.944673
1       10.34  1.66    Male     No  Sun  Dinner     3    16.054159
2       21.01  3.50    Male     No  Sun  Dinner     3    16.658734
3       23.68  3.31    Male     No  Sun  Dinner     2    13.978041
4       24.59  3.61  Female     No  Sun  Dinner     4    14.680765

In [10]:
tips['tip_percentage'] = (tips['tip']/tips['total_bill'])*100
tips.head(2)

total_bill   tip     sex smoker  day    time  size  tip_percentage
0       16.99  1.01  Female     No  Sun  Dinner     2        5.944673
1       10.34  1.66    Male     No  Sun  Dinner     3       16.054159

In [11]:
tips = tips.assign(tip_percent = round(tips['tip']/tips['total_bill'], 3))
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
0       16.99  1.01  Female     No  Sun  Dinner     2        5.944673   
1       10.34  1.66    Male     No  Sun  Dinner     3       16.054159   
2       21.01  3.50    Male     No  Sun  Dinner     3       16.658734   
3       23.68  3.31    Male     No  Sun  Dinner     2       13.978041   
4       24.59  3.61  Female     No  Sun  Dinner     4       14.680765   

   tip_percent  
0        0.059  
1        0.161  
2        0.167  
3        0.140  
4        0.147

<a id='where'></a>
## Sql where 
- Select with where caluse
- Select with where clause and it has 'and' condition
    - & vs and
- Select with where clause and it has 'or' condition    


In [12]:
'''
select total_bill, tip, time from tips where time='Dinner' limit 10
'''

tips[tips['time'] == 'Dinner'][['total_bill', 'tip', 'time']].head(10)

total_bill   tip    time
0       16.99  1.01  Dinner
1       10.34  1.66  Dinner
2       21.01  3.50  Dinner
3       23.68  3.31  Dinner
4       24.59  3.61  Dinner
5       25.29  4.71  Dinner
6        8.77  2.00  Dinner
7       26.88  3.12  Dinner
8       15.04  1.96  Dinner
9       14.78  3.23  Dinner

In [13]:
# this will return series of boolean. Truw when condion is met
is_dinner = tips['time'] == 'Dinner'

In [14]:
is_dinner.value_counts()

True     176
False     68
Name: time, dtype: int64

In [15]:
tips[is_dinner][['total_bill', 'tip', 'time']]

total_bill   tip    time
0         16.99  1.01  Dinner
1         10.34  1.66  Dinner
2         21.01  3.50  Dinner
3         23.68  3.31  Dinner
4         24.59  3.61  Dinner
..          ...   ...     ...
239       29.03  5.92  Dinner
240       27.18  2.00  Dinner
241       22.67  2.00  Dinner
242       17.82  1.75  Dinner
243       18.78  3.00  Dinner

[176 rows x 3 columns]

<p style ="color: rgb(0,120,120)"> So both above methods are similar. in the second one we can verify using **value_counts** on series that 176 records in tips data are for dinner time. 
<br>

**value_counts** gives tabulation of data</p>

In [16]:
 tips['tip'] > 5

0      False
1      False
2      False
3      False
4      False
       ...  
239     True
240    False
241    False
242    False
243    False
Name: tip, Length: 244, dtype: bool

In [17]:
tips['time'] == 'Dinner'

0      True
1      True
2      True
3      True
4      True
       ... 
239    True
240    True
241    True
242    True
243    True
Name: time, Length: 244, dtype: bool

In [18]:
'''
Select total_bill, tip, time from tips where time ='Dinner' and tip > 5 limit 10
'''

((tips['time'] == 'Dinner') & (tips['tip'] > 5)).value_counts()

False    229
True      15
dtype: int64

<font color='rgb(200,0, 0)'> 
In the above code two things are to be noted, 
- the two conditions **must** be in their own parantheses 
- notice usage of **&** ; & is used for bitwise comparison, it will return cases where both are true. If we have used 'and' it'll mean all or none
</font>
    

In [19]:
tips[(tips['time'] == 'Dinner') & (tips['tip'] > 5)][['total_bill', 'tip', 'time']].head(10)

total_bill    tip    time
23        39.42   7.58  Dinner
44        30.40   5.60  Dinner
47        32.40   6.00  Dinner
52        34.81   5.20  Dinner
59        48.27   6.73  Dinner
116       29.93   5.07  Dinner
155       29.85   5.14  Dinner
170       50.81  10.00  Dinner
172        7.25   5.15  Dinner
181       23.33   5.65  Dinner

In [20]:
'''
Select total_bill, tip, time from tips where total_bill > 50.0 or tip > 5 limit 10
'''

tips[(tips['total_bill'] > 50)  | (tips['tip'] > 5)][['total_bill', 'tip', 'time']].head(10) 

total_bill   tip    time
23        39.42  7.58  Dinner
44        30.40  5.60  Dinner
47        32.40  6.00  Dinner
52        34.81  5.20  Dinner
59        48.27  6.73  Dinner
85        34.83  5.17   Lunch
88        24.71  5.85   Lunch
116       29.93  5.07  Dinner
141       34.30  6.70   Lunch
155       29.85  5.14  Dinner

<font color='rgb(200,0, 0)'> 
Notice usage of **|** ; | is used for bitwise comparison, it will return cases where either of the condition is true. 
</font>

<a id='null'></a>
## IS NULL and NOT NULL

In [21]:
df = pd.DataFrame({
    'col_1': ['a', 'b', np.nan, 'd'],
    'col_2': [1, np.nan, np.nan, 100]
})
df

col_1  col_2
0     a    1.0
1     b    NaN
2   NaN    NaN
3     d  100.0

In [22]:
'''
Select * from df where col_2 = Null
''' 
df[df['col_2'].isna()]

col_1  col_2
1     b    NaN
2   NaN    NaN

In [23]:
'''
Select * from df where col_2 IS NOT Null
'''  
df[df['col_2'].notna()]

col_1  col_2
0     a    1.0
3     d  100.0

<a id='in'></a>
## Sql IN

In [24]:
'''
Select * from tips where tip in (2,3,5)
'''

tips[tips['tip'].isin([2,3,5])]



total_bill  tip     sex smoker   day    time  size  tip_percentage  \
6          8.77  2.0    Male     No   Sun  Dinner     2       22.805017   
11        35.26  5.0  Female     No   Sun  Dinner     4       14.180374   
13        18.43  3.0    Male     No   Sun  Dinner     4       16.277808   
26        13.37  2.0    Male     No   Sat  Dinner     2       14.958863   
27        12.69  2.0    Male     No   Sat  Dinner     2       15.760441   
..          ...  ...     ...    ...   ...     ...   ...             ...   
231       15.69  3.0    Male    Yes   Sat  Dinner     3       19.120459   
234       15.53  3.0    Male    Yes   Sat  Dinner     2       19.317450   
240       27.18  2.0  Female    Yes   Sat  Dinner     2        7.358352   
241       22.67  2.0    Male    Yes   Sat  Dinner     2        8.822232   
243       18.78  3.0  Female     No  Thur  Dinner     2       15.974441   

     tip_percent  
6          0.228  
11         0.142  
13         0.163  
26         0.150  
27         0.158  
..           ...  
231        0.191  
234        0.193  
240        0.074  
241        0.088  
243        0.160  

[66 rows x 9 columns]

<a id='groupby'></a>

## Sql Groupby

- count() vs size()
- applying multiple functions in group by 
    - agg() vs apply()
- group by multiple columns    

In [25]:
'''
Select count(*), sex from tips groupby sex
'''

tips.groupby(by = 'sex').size()

sex
Female     87
Male      157
dtype: int64

In [26]:
tips.groupby(by = 'sex').count()

total_bill  tip  smoker  day  time  size  tip_percentage  tip_percent
sex                                                                          
Female          87   87      87   87    87    87              87           87
Male           157  157     157  157   157   157             157          157

In [27]:
tips['sex'].value_counts()

Male      157
Female     87
Name: sex, dtype: int64

<font color='rgb(200,0, 0)'> 
Notice in pandas we use **size** that is sql equivalent. The count() function is applied to each column not the whole data
</font>

<br>

**count()** can be applied to individual column

In [28]:
tips.groupby(by = 'sex')['sex'].count()

sex
Female     87
Male      157
Name: sex, dtype: int64

**how tip amount differs by day of the week. In other words, how many tips you got on each day and what was avergae of tips for that date.**

In [29]:
tips.groupby(by='day').size()

day
Fri     19
Sat     87
Sun     76
Thur    62
dtype: int64

In [30]:
'''
Select day , count(tip), avg(tip) from tips groupby day
'''
#size and count both will work below because we have selected one column 'tip'
tips.groupby('day')['tip'].agg({'size', 'mean'})

size      mean
day                 
Fri     19  2.734737
Sat     87  2.993103
Sun     76  3.255132
Thur    62  2.771452

In [31]:
tips.groupby('day').agg({'tip':[np.size, np.mean]})

tip          
      size      mean
day                 
Fri   19.0  2.734737
Sat   87.0  2.993103
Sun   76.0  3.255132
Thur  62.0  2.771452

In [32]:
tips.groupby('day')['tip'].apply(np.mean)

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [33]:
'''Select day, count(tip), avg(tip) from tips groupby smoker, day'''

tips.groupby(['smoker', 'day']).agg({'tip': [np.size, np.mean]})

tip          
             size      mean
smoker day                 
No     Fri    4.0  2.812500
       Sat   45.0  3.102889
       Sun   57.0  3.167895
       Thur  45.0  2.673778
Yes    Fri   15.0  2.714000
       Sat   42.0  2.875476
       Sun   19.0  3.516842
       Thur  17.0  3.030000

<a id='having'></a>
## Sql Having
- filter 
- transform


In [34]:
'''
Select day, avg(total_bill) from tips groupby day having avg(total_bill) > 21
'''  

'\nSelect day, avg(total_bill) from tips groupby day having avg(total_bill) > 21\n'

In [35]:
tips_avg_bill = tips.groupby('day').agg({'total_bill': np.mean})
tips_avg_bill

total_bill
day             
Fri    17.151579
Sat    20.441379
Sun    21.410000
Thur   17.682742

In [36]:
tips_avg_bill[tips_avg_bill['total_bill'] > 21]

total_bill
day            
Sun       21.41

<font color='rgb(200,0, 0)'> 
If we want to return the **original rows** where avg bill of a day was > 21 then we use filter    
</font>    

In [37]:
tips.groupby(by='day').filter(lambda x : x['total_bill'].mean() > 21) 

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
0         16.99  1.01  Female     No  Sun  Dinner     2        5.944673   
1         10.34  1.66    Male     No  Sun  Dinner     3       16.054159   
2         21.01  3.50    Male     No  Sun  Dinner     3       16.658734   
3         23.68  3.31    Male     No  Sun  Dinner     2       13.978041   
4         24.59  3.61  Female     No  Sun  Dinner     4       14.680765   
..          ...   ...     ...    ...  ...     ...   ...             ...   
186       20.90  3.50  Female    Yes  Sun  Dinner     3       16.746411   
187       30.46  2.00    Male    Yes  Sun  Dinner     5        6.565988   
188       18.15  3.50  Female    Yes  Sun  Dinner     3       19.283747   
189       23.10  4.00    Male    Yes  Sun  Dinner     3       17.316017   
190       15.69  1.50    Male    Yes  Sun  Dinner     2        9.560229   

     tip_percent  
0          0.059  
1          0.161  
2          0.167  
3          0.140  
4          0.147  
..           ...  
186        0.167  
187        0.066  
188        0.193  
189        0.173  
190        0.096  

[76 rows x 9 columns]

<font color='rgb(200,0, 0)'> 
In the above result we got 76 rows because there are 76 rows for sunday and that can be validated by below code
</font>  

In [38]:
(tips['day'] == 'Sun').value_counts()

False    168
True      76
Name: day, dtype: int64

<font color='rgb(200,0, 0)'> 
Let's say we want to get the average bill of everyday for each original row. to do this we can use **transform**. It'll return the same number of rows but will modifiy the column on which transform is applied.
</font>  

In [39]:
'''This gives the average of grouped by day'''
tips.groupby('day')['total_bill'].apply(np.mean)

day
Fri     17.151579
Sat     20.441379
Sun     21.410000
Thur    17.682742
Name: total_bill, dtype: float64

In [40]:
'''This gives the avg of each day with original row, 
this can be used add an avg_bill_of_day column with original dataset'''

tips.groupby(by='day')['total_bill'].transform(lambda x: round(x.mean(), 2))

0      21.41
1      21.41
2      21.41
3      21.41
4      21.41
       ...  
239    20.44
240    20.44
241    20.44
242    20.44
243    17.68
Name: total_bill, Length: 244, dtype: float64

<a id='join'></a>

## Sql Join
Two options in pandas : join() , merge()
<br><br>
- **join()** by default joins on indices
- merge performs **inner join** by default    

In [41]:
# some sample data
df1 = pd.DataFrame({
        'key':['A', 'B', 'C', 'D', 'E', 'F'],
        'value': np.random.randn(6)
        })

df2 = pd.DataFrame({
        'key':['B', 'B', 'A', 'G', 'E', 'H'],
        'value': np.random.randn(6)
        })
print(df1)

df2

  key     value
0   A  1.336973
1   B -0.481101
2   C -0.654034
3   D -0.381398
4   E  0.388406
5   F  0.281368


key     value
0   B -1.114107
1   B -0.411327
2   A  3.144336
3   G -1.100962
4   E  0.067046
5   H -0.745556

### Inner Join

In [42]:
'''
Select * from df1, df2 where df1.key = df2.key

OR 

Select * from df1 inner join df2 on df1.key = df2.key

'''

pd.merge(df1, df2, on='key')

key   value_x   value_y
0   A  1.336973  3.144336
1   B -0.481101 -1.114107
2   B -0.481101 -0.411327
3   E  0.388406  0.067046

### Left Outer Join

In [43]:
'''
Select * from df1 left join df2 on df1.key = df2.key
'''
pd.merge(df1, df2, on='key', how='left')

key   value_x   value_y
0   A  1.336973  3.144336
1   B -0.481101 -1.114107
2   B -0.481101 -0.411327
3   C -0.654034       NaN
4   D -0.381398       NaN
5   E  0.388406  0.067046
6   F  0.281368       NaN

### Right Outer Join

In [44]:
'''Select * from df1 right join df2 on df1.key = df2.key'''
pd.merge(df1, df2, on='key', how='right')

key   value_x   value_y
0   A  1.336973  3.144336
1   B -0.481101 -1.114107
2   B -0.481101 -0.411327
3   E  0.388406  0.067046
4   G       NaN -1.100962
5   H       NaN -0.745556

### Full/Outer Join

In [45]:
'''Select * from df1 outer join df2 on df1.key = df2.key '''
pd.merge(df1, df2, on='key', how='outer')

key   value_x   value_y
0   A  1.336973  3.144336
1   B -0.481101 -1.114107
2   B -0.481101 -0.411327
3   C -0.654034       NaN
4   D -0.381398       NaN
5   E  0.388406  0.067046
6   F  0.281368       NaN
7   G       NaN -1.100962
8   H       NaN -0.745556

### Joining on Multiple keys

In [46]:
df3 = pd.DataFrame({
        'key':['B', 'B', 'A', 'G', 'E', 'H'],
        'key_2': [10,20,30,40,50,60], 
        'value': np.random.randn(6)
        })

df4 = pd.DataFrame({
        'key':['B', 'B', 'A', 'G', 'E', 'H'],
        'key_2': [10,70,30,80,20,60], 
        'value': np.random.randn(6)
        })

print(df3)
df4

  key  key_2     value
0   B     10 -0.402497
1   B     20  0.848472
2   A     30  0.238164
3   G     40 -1.360144
4   E     50 -0.623638
5   H     60  0.238283


key  key_2     value
0   B     10  0.355848
1   B     70 -0.161952
2   A     30 -1.026339
3   G     80  1.287335
4   E     20 -2.836549
5   H     60 -0.190881

In [47]:
pd.merge(df3, df4, left_on=['key', 'key_2'], right_on=['key', 'key_2'])

key  key_2   value_x   value_y
0   B     10 -0.402497  0.355848
1   A     30  0.238164 -1.026339
2   H     60  0.238283 -0.190881

In [48]:
# if column names are same the following syntax also works
pd.merge(df3, df4, on=['key', 'key_2'])

key  key_2   value_x   value_y
0   B     10 -0.402497  0.355848
1   A     30  0.238164 -1.026339
2   H     60  0.238283 -0.190881

<a id='update'></a>
## Sql Update

update syntax is:
>df.loc[filter_criteria_f_row_selection, column_to_be_updated] = new_val

In [49]:
'''Lets add some dummy columns to be updated'''
tips['desired_tip'] = round( tips['total_bill']*.25, 2)
tips['flag_tips'] = False
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
0       16.99  1.01  Female     No  Sun  Dinner     2        5.944673   
1       10.34  1.66    Male     No  Sun  Dinner     3       16.054159   
2       21.01  3.50    Male     No  Sun  Dinner     3       16.658734   
3       23.68  3.31    Male     No  Sun  Dinner     2       13.978041   
4       24.59  3.61  Female     No  Sun  Dinner     4       14.680765   

   tip_percent  desired_tip  flag_tips  
0        0.059         4.25      False  
1        0.161         2.58      False  
2        0.167         5.25      False  
3        0.140         5.92      False  
4        0.147         6.15      False

In [50]:
'''
set flag_tips = True where difference between desired tip  and actual tip is more than 3

SQL -->

update flag_tips = True 
where abs( tip - desired_tip) > 3
'''
update_cond = np.abs(tips['tip'] - tips['desired_tip']) > 3
tips.loc[update_cond, 'flag_tips'] = True


In [51]:
tips['flag_tips'].value_counts()

False    196
True      48
Name: flag_tips, dtype: int64

<font color='rgb(200,0, 0)'> 
48 times tip is too less. needs to look at them ;)
</font>

<a id='delete'></a>
## Sql Delete
Syntax for delete is same as for update. <br> MAIN DIFFERENCE is 'filter criteria' should pick rows that should not be deleted and **assign** it to a new dataframe variable or update the same data frame variable

In [52]:
'''
Delete tips where tip_percentage > 30. # becase no one gives that much tip
'''

tips[tips['tip_percentage'] >30]

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
67         3.07  1.00  Female    Yes  Sat  Dinner     1       32.573290   
172        7.25  5.15    Male    Yes  Sun  Dinner     2       71.034483   
178        9.60  4.00  Female    Yes  Sun  Dinner     2       41.666667   

     tip_percent  desired_tip  flag_tips  
67         0.326         0.77      False  
172        0.710         1.81       True  
178        0.417         2.40      False

In [53]:
#update
tips = tips[tips['tip_percentage'] < 30]

<a id="union"></a>
## Sql Union All and Union
- union all combines all records and can have duplicates
- union does not have duplicates

In [54]:
pd.concat([df3.iloc[:,:2], df4.iloc[:,:2]])


key  key_2
0   B     10
1   B     20
2   A     30
3   G     40
4   E     50
5   H     60
0   B     10
1   B     70
2   A     30
3   G     80
4   E     20
5   H     60

In [55]:
# sql union 
pd.concat([df3.iloc[:,:2], df4.iloc[:,:2]]).drop_duplicates()


key  key_2
0   B     10
1   B     20
2   A     30
3   G     40
4   E     50
5   H     60
1   B     70
3   G     80
4   E     20

<a id='windows_functions'></a>
## Sql Window Functions

### Select top 'n' rows per group

The following query is adding a row number to records by grouping them by day. Meaning, if the dataset has 10 records for friday then the row numbers for friday will strat from 1 to 10. 

>    select * from 
    (select t.*, row_number() over (partition by day order by total_bill) as rn
    from tips t ) where row_number < 3 
    
**partition by** clause divides the query's result into partitions
<br>
**row_numberber()** is a windows's function and is operated on each partition separately and recalculate for each partition
<br>
**cumcount()**  number each item in each group from 0 to the len(group) - 1
<br>
**query()** query the columns of dataframe with a boolean expression
<br>
**rank()** function assigns rank per partition, similar to cumcount

In [56]:
temp = tips.assign(row_num = tips.sort_values(by=['total_bill'], ascending=True).groupby('day').cumcount()+1)
temp.head()

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
0       16.99  1.01  Female     No  Sun  Dinner     2        5.944673   
1       10.34  1.66    Male     No  Sun  Dinner     3       16.054159   
2       21.01  3.50    Male     No  Sun  Dinner     3       16.658734   
3       23.68  3.31    Male     No  Sun  Dinner     2       13.978041   
4       24.59  3.61  Female     No  Sun  Dinner     4       14.680765   

   tip_percent  desired_tip  flag_tips  row_num  
0        0.059         4.25       True       26  
1        0.161         2.58      False        7  
2        0.167         5.25      False       41  
3        0.140         5.92      False       48  
4        0.147         6.15      False       52

In [57]:
temp.query("day == 'Fri' and row_num < 4")

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
92         5.75  1.00  Female    Yes  Fri  Dinner     2       17.391304   
222        8.58  1.92    Male    Yes  Fri   Lunch     1       22.377622   
226       10.09  2.00  Female    Yes  Fri   Lunch     2       19.821606   

     tip_percent  desired_tip  flag_tips  row_num  
92         0.174         1.44      False        1  
222        0.224         2.14      False        2  
226        0.198         2.52      False        3

In [58]:
temp.query("row_num < 3").sort_values(by=['day', 'row_num'])

total_bill   tip     sex smoker   day    time  size  tip_percentage  \
92         5.75  1.00  Female    Yes   Fri  Dinner     2       17.391304   
222        8.58  1.92    Male    Yes   Fri   Lunch     1       22.377622   
111        7.25  1.00  Female     No   Sat  Dinner     1       13.793103   
218        7.74  1.44    Male    Yes   Sat  Dinner     2       18.604651   
6          8.77  2.00    Male     No   Sun  Dinner     2       22.805017   
43         9.68  1.32    Male     No   Sun  Dinner     2       13.636364   
149        7.51  2.00    Male     No  Thur   Lunch     2       26.631158   
195        7.56  1.44    Male     No  Thur   Lunch     2       19.047619   

     tip_percent  desired_tip  flag_tips  row_num  
92         0.174         1.44      False        1  
222        0.224         2.14      False        2  
111        0.138         1.81      False        1  
218        0.186         1.94      False        2  
6          0.228         2.19      False        1  
43         0.136         2.42      False        2  
149        0.266         1.88      False        1  
195        0.190         1.89      False        2

<font color='rgb(200,0, 0)'> 
Do we need to sort before partitioning? </font> YES, answer will be different , see below code


In [59]:
tips.assign(rn=tips.groupby(by= 'day').cumcount()+1).query('rn < 3').sort_values(by=['day', 'rn'])

total_bill   tip     sex smoker   day    time  size  tip_percentage  \
90       28.97  3.00    Male    Yes   Fri  Dinner     2       10.355540   
91       22.49  3.50    Male     No   Fri  Dinner     2       15.562472   
19       20.65  3.35    Male     No   Sat  Dinner     3       16.222760   
20       17.92  4.08    Male     No   Sat  Dinner     2       22.767857   
0        16.99  1.01  Female     No   Sun  Dinner     2        5.944673   
1        10.34  1.66    Male     No   Sun  Dinner     3       16.054159   
77       27.20  4.00    Male     No  Thur   Lunch     4       14.705882   
78       22.76  3.00    Male     No  Thur   Lunch     2       13.181019   

    tip_percent  desired_tip  flag_tips  rn  
90        0.104         7.24       True   1  
91        0.156         5.62      False   2  
19        0.162         5.16      False   1  
20        0.228         4.48      False   2  
0         0.059         4.25       True   1  
1         0.161         2.58      False   2  
77        0.147         6.80      False   1  
78        0.132         5.69      False   2

In [60]:
'''Using query() to query tips dataframe'''
tips.query('(total_bill >20 & total_bill <25) & tip< 5 & sex==\'Female\'')

total_bill   tip     sex smoker   day    time  size  tip_percentage  \
4         24.59  3.61  Female     No   Sun  Dinner     4       14.680765   
21        20.29  2.75  Female     No   Sat  Dinner     2       13.553475   
33        20.69  2.45  Female     No   Sat  Dinner     4       11.841469   
94        22.75  3.25  Female     No   Fri  Dinner     2       14.285714   
103       22.42  3.48  Female    Yes   Sat  Dinner     2       15.521855   
104       20.92  4.08  Female     No   Sat  Dinner     2       19.502868   
119       24.08  2.92  Female     No  Thur   Lunch     4       12.126246   
131       20.27  2.83  Female     No  Thur   Lunch     2       13.961519   
186       20.90  3.50  Female    Yes   Sun  Dinner     3       16.746411   
229       22.12  2.88  Female    Yes   Sat  Dinner     2       13.019892   

     tip_percent  desired_tip  flag_tips  
4          0.147         6.15      False  
21         0.136         5.07      False  
33         0.118         5.17      False  
94         0.143         5.69      False  
103        0.155         5.60      False  
104        0.195         5.23      False  
119        0.121         6.02       True  
131        0.140         5.07      False  
186        0.167         5.22      False  
229        0.130         5.53      False

Using **rank()** to get row_number

In [61]:
tips.assign(rn = tips.groupby('day')['total_bill'].rank(method='first',ascending=False)).query('rn< 3').sort_values(by=['day', 'rn'])

total_bill    tip     sex smoker   day    time  size  tip_percentage  \
95        40.17   4.73    Male    Yes   Fri  Dinner     4       11.774956   
90        28.97   3.00    Male    Yes   Fri  Dinner     2       10.355540   
170       50.81  10.00    Male    Yes   Sat  Dinner     3       19.681165   
212       48.33   9.00    Male     No   Sat  Dinner     4       18.621974   
156       48.17   5.00    Male     No   Sun  Dinner     6       10.379905   
182       45.35   3.50    Male    Yes   Sun  Dinner     3        7.717751   
197       43.11   5.00  Female    Yes  Thur   Lunch     4       11.598237   
142       41.19   5.00    Male     No  Thur   Lunch     5       12.138869   

     tip_percent  desired_tip  flag_tips   rn  
95         0.118        10.04       True  1.0  
90         0.104         7.24       True  2.0  
170        0.197        12.70      False  1.0  
212        0.186        12.08       True  2.0  
156        0.104        12.04       True  1.0  
182        0.077        11.34       True  2.0  
197        0.116        10.78       True  1.0  
142        0.121        10.30       True  2.0

In [62]:
'''
Let’s find tips with (rank < 3) per gender group for (tips < 2)
'''

temp_tip = tips.assign(min_tip = tips[tips['tip'] < 2].groupby('sex')['tip'].rank(method='min'))
temp_tip.head()

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
0       16.99  1.01  Female     No  Sun  Dinner     2        5.944673   
1       10.34  1.66    Male     No  Sun  Dinner     3       16.054159   
2       21.01  3.50    Male     No  Sun  Dinner     3       16.658734   
3       23.68  3.31    Male     No  Sun  Dinner     2       13.978041   
4       24.59  3.61  Female     No  Sun  Dinner     4       14.680765   

   tip_percent  desired_tip  flag_tips  min_tip  
0        0.059         4.25       True      3.0  
1        0.161         2.58      False     20.0  
2        0.167         5.25      False      NaN  
3        0.140         5.92      False      NaN  
4        0.147         6.15      False      NaN

In [63]:
temp_tip.query('min_tip < 3')

total_bill   tip     sex smoker  day    time  size  tip_percentage  \
92         5.75  1.00  Female    Yes  Fri  Dinner     2       17.391304   
111        7.25  1.00  Female     No  Sat  Dinner     1       13.793103   
236       12.60  1.00    Male    Yes  Sat  Dinner     2        7.936508   
237       32.83  1.17    Male    Yes  Sat  Dinner     2        3.563814   

     tip_percent  desired_tip  flag_tips  min_tip  
92         0.174         1.44      False      1.0  
111        0.138         1.81      False      1.0  
236        0.079         3.15      False      1.0  
237        0.036         8.21       True      2.0

###  Top 'n' rows with an offset

- nlarhest () --> Return the first n rows ordered by columns in descending order
- nsmalles() reverse of nlargest()

In [64]:
tips.nlargest(10, 'tip')

total_bill    tip     sex smoker   day    time  size  tip_percentage  \
170       50.81  10.00    Male    Yes   Sat  Dinner     3       19.681165   
212       48.33   9.00    Male     No   Sat  Dinner     4       18.621974   
23        39.42   7.58    Male     No   Sat  Dinner     4       19.228818   
59        48.27   6.73    Male     No   Sat  Dinner     4       13.942407   
141       34.30   6.70    Male     No  Thur   Lunch     6       19.533528   
183       23.17   6.50    Male    Yes   Sun  Dinner     4       28.053517   
214       28.17   6.50  Female    Yes   Sat  Dinner     3       23.074192   
47        32.40   6.00    Male     No   Sun  Dinner     4       18.518519   
239       29.03   5.92    Male     No   Sat  Dinner     3       20.392697   
88        24.71   5.85    Male     No  Thur   Lunch     2       23.674626   

     tip_percent  desired_tip  flag_tips  
170        0.197        12.70      False  
212        0.186        12.08       True  
23         0.192         9.86      False  
59         0.139        12.07       True  
141        0.195         8.57      False  
183        0.281         5.79      False  
214        0.231         7.04      False  
47         0.185         8.10      False  
239        0.204         7.26      False  
88         0.237         6.18      False

In [65]:
tips.nlargest(10, 'tip').tail(5)

total_bill   tip     sex smoker   day    time  size  tip_percentage  \
183       23.17  6.50    Male    Yes   Sun  Dinner     4       28.053517   
214       28.17  6.50  Female    Yes   Sat  Dinner     3       23.074192   
47        32.40  6.00    Male     No   Sun  Dinner     4       18.518519   
239       29.03  5.92    Male     No   Sat  Dinner     3       20.392697   
88        24.71  5.85    Male     No  Thur   Lunch     2       23.674626   

     tip_percent  desired_tip  flag_tips  
183        0.281         5.79      False  
214        0.231         7.04      False  
47         0.185         8.10      False  
239        0.204         7.26      False  
88         0.237         6.18      False

In [66]:
tips.nsmallest(10, 'tip').tail(5)

total_bill   tip     sex smoker   day    time  size  tip_percentage  \
237       32.83  1.17    Male    Yes   Sat  Dinner     2        3.563814   
75        10.51  1.25    Male     No   Sat  Dinner     2       11.893435   
135        8.51  1.25  Female     No  Thur   Lunch     2       14.688602   
235       10.07  1.25    Male     No   Sat  Dinner     2       12.413108   
43         9.68  1.32    Male     No   Sun  Dinner     2       13.636364   

     tip_percent  desired_tip  flag_tips  
237        0.036         8.21       True  
75         0.119         2.63      False  
135        0.147         2.13      False  
235        0.124         2.52      False  
43         0.136         2.42      False